### PR TITLE
🪪 Adds a more fine-grained profiling context

### PR DESCRIPTION
**Pull request by edbeeching**
# What does this PR do?

- Adds a more fine-grained profiling context
- Adds profiling contexts around vllm.generate and each reward function

### CODE REVIEW

File: trl/extras/profiling.py [hunk 1]
```diff
@@ -22,20 +23,29 @@
     import wandb
 
 
+@contextlib.contextmanager
+def profiling_context(instance, name):
```
Comment (qgallouedec) [inline trl/extras/profiling.py:27]:
```suggestion
def profiling_context(trainer, name):
```

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -786,23 +787,30 @@ def _generate_and_score_completions(
...
-                    texts = [apply_chat_template(x, reward_processing_class)["text"] for x in messages]
-                else:
-                    texts = [p + c for p, c in zip(prompts, completions)]
-                reward_inputs = reward_processing_class(
-                    texts, return_tensors="pt", padding=True, padding_side="right", add_special_tokens=False
-                )
-                reward_inputs = super()._prepare_inputs(reward_inputs)
-                with torch.inference_mode():
-                    rewards_per_func[:, i] = reward_func(**reward_inputs).logits[:, 0]  # Shape (B*G,)
+                reward_func_name = f"reward {reward_func.config._name_or_path.split('/')[-1]}"
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:790]: I think it should be

```suggestion
                reward_func_name = f"reward/ {reward_func.config._name_or_path.split('/')[-1]}"
```

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:790]: there is a white space that I didn't see actually

**Review by qgallouedec — APPROVED** [review]
Very useful!
